pkg/controller/ads/dns.go

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ads
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+	"slices"
+	"sync"
+	"time"
+
+	clusterv3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	endpointv3 "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	core_v2 "kmesh.net/kmesh/api/v2/core"
+	"kmesh.net/kmesh/pkg/dns"
+)
+
+// adsDnsResolver is DNS resolver of Kernel Native
+type dnsController struct {
+	Clusters    chan []*clusterv3.Cluster
+	cache       *AdsCache
+	dnsResolver *dns.DNSResolver
+	// Store the copy of pendingResolveDomain.
+	clusterCache map[string]*pendingResolveDomain
+	// store all pending hostnames in the clusters
+	pendingClusterinfo map[string]clusterInfo
+	sync.RWMutex
+}
+
+// pending resolve domain info of Kennel-Native Mode,
+// domain name is used for dns resolution
+// cluster is used for create the apicluster
+type pendingResolveDomain struct {
+	DomainName  string
+	Clusters    []*clusterv3.Cluster
+	RefreshRate time.Duration
+}
+
+type clusterInfo struct {
+	info map[string]hostInfo
+}
+
+type hostInfo struct {
+	addresses []string
+}
+
+func NewDnsResolver(adsCache *AdsCache) (*dnsController, error) {
+	resolver, err := dns.NewDNSResolver()
+	if err != nil {
+		return nil, err
+	}
+	return &dnsController{
+		Clusters:           make(chan []*clusterv3.Cluster),
+		cache:              adsCache,
+		dnsResolver:        resolver,
+		clusterCache:       make(map[string]*pendingResolveDomain),
+		pendingClusterinfo: make(map[string]clusterInfo),
+	}, nil
+}
+
+func (r *dnsController) StartKernelNativeDnsController(stopCh <-chan struct{}) {
+	// start dns resolver
+	go r.dnsResolver.StartDnsResolver(stopCh)
+	go r.refreshAdsWorker(stopCh)
+	go r.startDnsController()
+	go func() {
+		<-stopCh
+		close(r.Clusters)
+	}()
+}
+
+func (r *dnsController) startDnsController() {
+	for clusters := range r.Clusters {
+		go func(clusters []*clusterv3.Cluster) {
+			r.resolveDomains(clusters)
```

Comment:
I think this is not safe running async in separate routines,

think about this case:

t0:    a resolveDomains thread

t1:   another resolveDomains

If t1 is run before t0, even some step.The state machine would be broken